### PR TITLE
Allow pickling

### PR DESCRIPTION
**Pull request by lbluque**
Addresses #386 

* Implement `__reduce__` in cython extension classes to allow pickling.
* add unit tests for pickling most objects.

## Checklist

- [X] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
      Run [pycodestyle](https://pycodestyle.readthedocs.io/en/latest/) and [flake8](http://flake8.pycqa.org/en/latest/)
      on your local machine.
- [X] Docstrings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
      Run [pydocstyle](http://www.pydocstyle.org/en/2.1.1/index.html) on your code.
- [X] Tests have been added for any new functionality or bug fixes.
- [X] All existing tests pass.

### CODE REVIEW

**Comment by lbluque** [comment]
@qchempku2017 pls have a look

File: tests/utils.py [hunk 1]
```diff
@@ -41,7 +42,19 @@ def assert_msonable(obj, skip_keys=None, test_if_subclass=True):
...
         assert d1[key] == d2[key]
-    _ = json.loads(obj.to_json(), cls=MontyDecoder)
+
+    try:
+        _ = json.loads(obj.to_json(), cls=MontyDecoder)
+    except Exception as e:
+        raise AssertionError(e)
+
+
+def assert_pickles(obj):
```
Comment (qchempku2017) [inline tests/utils.py:52]: Could you please test pickle load too?

Comment (qchempku2017) [reply tests/utils.py:52]: If comparison is hard, there is no need to check comparison. But it should be better if we can test whether pickled objects can be reloaded.

Comment (lbluque) [reply tests/utils.py:52]: I added pickle load to the tests.

testing equality is a bit hard unless we implement an __eq__ for all the classes we are testing pickling. Instead, I implemented comparing MSONable dictionaries if possible, and a weak sanity check for other objects. I think that is enough.

Let me know what you think.

Comment (qchempku2017) [reply tests/utils.py:52]: Thanks, that's great!

**Review by qchempku2017 — APPROVED** [review]

**Pull request merged**